keps/sig-auth/0034-20190123-harden-default-discovery-bindings.md

```diff
@@ -0,0 +1,71 @@
+
+# Harden Default RBAC Discovery ClusterRole(Binding)s
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a KEP and for highlighting any additional information provided beyond the standard KEP template.
+[Tools for generating][] a table of contents from markdown are available.
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+The aim of this change is to remove the `system:unauthenticated` subjects from the `system:discovery` and `system:basic-user` ClusterRoleBindings, while preserving unauthenticated access to a genuinely non-sensitive subset of the current APIs (e.g. `GET /healthz`, `GET /version`). This will improve the default privacy and security posture for new clusters.
+
+## Motivation
+
+One of the work items resulting from the [CVE-2018-1002105](https://github.com/kubernetes/kubernetes/issues/71411) post-mortem was to [investigate hardening the default RBAC discovery ClusterRoleBindings](https://github.com/kubernetes/kubernetes/issues/72115) (i.e. `system:discovery` and `system:basic-user`) to limit potential avenues for similar attack. Additionally, the fact that API extensions are exposed by the default discovery bindings is surprising to some and represents a potential privacy concern (e.g. `/apis/self-flying-cars.unicorn.vc/v1/`).
+
+### Goals
+
+Limit the set of APIs which allow for unauthenticated access by default.
+
+### Non-Goals
+
+Disabling unauthenticated and/or anonymous access to all APIs by default.
+
+## Proposal
+
+* Remove the `GET /healthz`, `/version` and `/version/` PolicyRules from the default [`system:discovery` ClusterRole](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L198-L208)
+* Remove the `system:unauthenticated` subject group from the default [`system:discovery` and `system:basic-user` ClusterRoleBindings](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L526-L527)
+* Create a new default ClusterRole which prominently, explicitly documents its intent to freely disclose information (e.g. `system:public-info-viewer`) and contains the `GET /healthz`, `/version` and `/version/` PolicyRules removed from `system:discovery`
+* Create a new default ClusterRoleBinding for the new ClusterRole, which grants the `system:(un)authenticated` subject groups access.
+
+### Risks and Mitigations
+
+1. If a user upgrades a cluster which has already limited access to the `GET /healthz`, `/version` or `/version/` APIs, auto-reconciliation will re-enable universal access via `system:public-info-viewer`
```

Comment:
> If a user upgrades a cluster which has already limited access to the `GET /healthz`, `/version` or `/version/` APIs, auto-reconciliation will re-enable universal access via `system:public-info-viewer`

we had a similar scenario when we introduced aggregated cluster roles and moved the source of the admin/edit/view roles out to `aggregate-to*` roles. in that case, we introduced a migration step to the API server RBAC reconciliation that copied out the existing role content to the new separate roles (https://github.com/kubernetes/kubernetes/blob/8b98e802eddb9f478ff7d991a2f72f60c165388a/pkg/registry/rbac/rest/storage_rbac.go#L162-L167)

A similar approach here could work. Before reconciling, if a `system:discovery` clusterrolebinding exists and the new `system:public-info-viewer` clusterrolebinding does not exist, copy existing subjects and the existing reconciliation annotation value from the `system:discovery` clusterrolebinding to form the new `system:public-info-viewer` clusterrolebinding. That preserves any existing customization previously applied to the clusterrolebinding.

Reply:
done